src/chkstat.cpp

```diff
@@ -612,8 +639,83 @@ bool Chkstat::parseProfile(const std::string &path)
             continue;
         }
 
-        const auto &entry = addProfileEntry(location, user, group, mode_int);
-        last_key = entry.file;
+        std::vector<std::string> expansions;
+        if (!expandProfilePaths(location, expansions) )
+        {
+            badProfileLine(path, linenr, "bad variable expansions");
+            continue;
+        }
+
+        for (const auto &exp_path: expansions)
+        {
+            addProfileEntry(exp_path, user, group, mode_int);
+            // remember the most recently added entries to allow potential '+'
+            // capability lines to be applied to them.
+            active_keys.push_back(exp_path);
+        }
+    }
+
+    return true;
+}
+
+bool Chkstat::expandProfilePaths(const std::string &path, std::vector<std::string> &expansions)
+{
+    std::stringstream ss;
+    ss.str(path);
+    std::string part;
+
+    // the initial entry
+    expansions.clear();
+    expansions.push_back("");
+
+    // process each path component.
+    //
+    // we support variables only as individual
+    // path components i.e. something like %{myvar}stuff/suffix is not
+    // allowed, only %{myvar}/suffix.
+    //
+    // multiple variable components in the same path are supported
+    while (std::getline(ss, part, '/'))
+    {
+        if (hasPrefix(part, "%{") && hasSuffix(part, "}"))
+        {
+            const auto variable = part.substr(2, part.length() - 3);
+            auto it = m_variable_expansions.find(variable);
+
+            if (it == m_variable_expansions.end())
+            {
+                expansions.clear();
+                std::cerr << "Undeclared variable %{" << variable << "} encountered." << std::endl;
+                return false;
+            }
+
+            // now we need to create additional entries for each possible
+            // value of the variable
+            std::vector<std::string> new_expansions;
+
+            for (const auto &element: expansions)
+            {
+                for (const auto &var_value: it->second)
+                {
+                    std::string sep = var_value[0] == '/' ? "" : "/";
```

Comment:
Can `var_value` ever be empty here? Maybe we can enforce all variable values starting with a slash when reading them in?

Reply:
Since the parser currently splits words at whitespace boundaries for the variable values no value should ever be empty. Even then we would only access the string terminator here.

I thought about requiring them starting with a slash, too. I fear, however, that this could limit the usability of variables or at least make them strange to use if you ever happen to want it to contain only a single path component.

Reply:
Maybe I should normalize the variables by removing leading slashes and also duplicate ones and trailing ones. How about that?

Reply:
Yes, that makes sense.